File: autocomplete/Producer.php
<?php

/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

namespace RocketMQ;

/**
 * 消息生产者
 * Class Producer
 * @package RocketMQ
 */
class Producer
{

	/**
	 * Producer constructor.
	 * @param string $groupId 
	 */
	public function __construct(string $groupId)
	{
	}

	/**
	 * 启动
	 */
	public function start()
	{
	}

	public function shutdown()
	{
	}

	/**
	 * @return version
	 */
	public function version()
	{
	}

	/**
	 * 设置名字服务链接地址
	 * @param string $nameserver
	 */
	public function setNameServerAddress(string $nameserver)
	{
	}

	/**
	 * 设置名字服务domain
	 * @param string domain$
	 */
	public function setNameServerDomain(string $domain)
	{
	}

	/**
	 * 设置分组名
	 * @param string $groupName
	 */
	public function setGroupName(string $groupName)
	{
	}


	/**
	 * 设置实例名
	 * @param string $instanceName  实例名
	 */
	public function setInstanceName(string $instanceName)
	{
	}


	/**
	 * 阿里云商用版本需要使用这个方法链接，开源版本不需要管
	 * @param string $accessKey
	 * @param string $secretKey
	 * @param string $autChannel
	 */
	public function setSessionCredentials(string $accessKey, string $secretKey, string $autChannel)
	{
	}

	/**
	 *  设置日志路径
	 */
	public function setLogPath(string $logPath)
	{
	}

	/**
	 * 设置日志生成规则
	 * @param $fileNum //保留日志数目
	 * @param $perFileSize //每个日志多大
	 */
	public function setLogFileNumAndSize($fileNum, $perFileSize)
	{
	}

	/**
	 * 设置日志等级
	 * @param int $inputLevel
	 */
	public function setLogLevel(int $inputLevel)
	{
	}

	/**
	 * 发送消息超时时间
	 * @param $sendMsgTimeout 单位是毫秒  默认3000毫秒
	 */
	public function setSendMsgTimeout($sendMsgTimeout)
	{
	}

	/**
	 * 设置消息压缩等级   level = [-1, 9]
	 * @param $compressLevel //默认是5
	 */
	public function setCompressLevel($compressLevel)
	{
	}

	/**
	 * 设置消息长度最大限制
	 * @param $messageSize
	 */
	public function setMaxMessageSize($messageSize)
	{
	}

	/**
	 * @param int traceModel
	 * @return int
	 * @see TraceModel
	 */
	public function setMessageTrace($traceModel)
	{
	}

	/**
	 * @param Message $message
	 * @return int
	 */
	public function setMessageSync($message)
	{

	}

	/**
	 * Summary of setMessageAsync
	 * @param Message $message
	 * @param callable $successCallback function(SendResult $sendresult, Message $message, $userData) 
	 * @param callable $exceptionCallback function(MQException $exception, Message $message, $userData) 
	 * @param mixed $userData
	 * @return void
	 */
	public function sendMessageAsync($message, $successCallback, $exceptionCallback, $userData)
	{

	}

	/**
	 * @param Message $message
	 * @return int
	 */
	public function sendMessageOneway($message)
	{

	}

	/**
	 * @param \RocketMQ\Message $message
	 * @param callable $selector function (int $size, Message $message, $userData)
	 * @param mixed $userData
	 * @return void
	 */
	public function sendMessageOnewayOrderly(Message $message,callable $selector, $userData)
	{

	}

	/**
	 * @param \RocketMQ\Message $message
	 * @param callable $selector function (int $size, Message $message, $userData)
	 * @param mixed $userData
	 * @param int $retryTimes
	 * @param \RocketMQ\SendResult $sendResult
	 * @return void
	 */
	public function sendMessageOrderly(Message $message, callable $selector, $userData, int $retryTimes, SendResult $sendResult)
	{

	}

	/**
	 * @param \RocketMQ\Message $message
	 * @param string $shardingKey
	 * @param \RocketMQ\SendResult $result
	 * @return void
	 */
	public function sendMessageOrderlyByShardingKey(Message $message, string $shardingKey, SendResult $result)
	{

	}

	/**
	 * @param \RocketMQ\Message $message
	 * @param callable $callback function(Message $message, $userData)
	 * @param mixed $userData
	 * @param \RocketMQ\SendResult $sendResult
	 * @return void
	 */
	public function sendMessageTranscation(Message $message, callable $callback, $userData, SendResult $sendResult)
	{

	}

	/**
	 * @param \RocketMQ\Message $message
	 * @param callable $callback function(Message $message, $userData)
	 * @param mixed $userData
	 * @param callable $successCallback function (SendResult $sendResult)
	 * @param callable $exceptionCallback function(MQException $mqException)
	 * @return void
	 */
	public function sendMessageOrderlyAsync(Message $message, callable $callback, $userData, callable $successCallback, callable $exceptionCallback)
	{

	}
}

File: src/producer.cc
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

#include "producer.h"
#include "message_queue.h"
#include "session_credentials.h"
#include "send_result.h"

void Producer::__construct(Php::Parameters &param){
	std::string groupName = param[0];
	this->producer = new rocketmq::DefaultMQProducer(groupName);
}

void Producer::setInstanceName(Php::Parameters &param){
	std::string instanceName = param[0];
	this->producer->setInstanceName(instanceName);
}

Php::Value Producer::getInstanceName(){
	return this->producer->getInstanceName();
}


void Producer::setGroupName(Php::Parameters &param){
	std::string groupName = param[0];
	this->producer->setGroupName(groupName);
}

Php::Value Producer::getGroupName(){
	return this->producer->getGroupName();
}

void Producer::setNamesrvAddr(Php::Parameters &param){
	std::string nameserver =  param[0];
	this->producer->setNamesrvAddr(nameserver);
}

void Producer::start(){
	this->producer->start();
}

void Producer::shutdown(){
	this->producer->shutdown();
}

Php::Value Producer::send(Message *message){
	rocketmq::SendResult sr = this->producer->send(message->getMQMessage());
	Php::Value pv(Php::Object(SEND_RESULT_CLASS_NAME, new SendResult(sr)));
	return pv;
}

Php::Value Producer::send(Message *message, bool bSelectActiveBroker ){
	rocketmq::SendResult sr = this->producer->send(message->getMQMessage(), bSelectActiveBroker);
	Php::Value pv(Php::Object(SEND_RESULT_CLASS_NAME, new SendResult(sr)));
	return pv;
}

Php::Value Producer::send(Message *message, Php::Value pvMessageQueue){
	MessageQueue* messageQueue = (MessageQueue*)pvMessageQueue.implementation();
	rocketmq::SendResult sr = this->producer->send(message->getMQMessage(), messageQueue->getInstance());
	Php::Value pv(Php::Object(SEND_RESULT_CLASS_NAME, new SendResult(sr)));
	return pv;
}

Php::Value Producer::send(std::vector<rocketmq::MQMessage> messages){
	rocketmq::SendResult sr = this->producer->send(messages);
	Php::Value pv(Php::Object(SEND_RESULT_CLASS_NAME, new SendResult(sr)));
	return pv;
}

Php::Value Producer::send(std::vector<rocketmq::MQMessage> messages,  Php::Value pvMessageQueue){
	MessageQueue* messageQueue = (MessageQueue*)pvMessageQueue.implementation();
	rocketmq::SendResult sr = this->producer->send(messages, messageQueue->getInstance());
	Php::Value pv(Php::Object(SEND_RESULT_CLASS_NAME, new SendResult(sr)));
	return pv;
}

// SendResult send(MQMessage& msg, const MQMessageQueue& mq);
Php::Value Producer::send(Php::Parameters &params){

	if (params[0].isArray()) {
		std::vector<rocketmq::MQMessage> messages;
		std::vector<Php::Value> vals= params[0].vectorValue<Php::Value>();

		for (unsigned int i = 0; i < vals.size(); i ++){
			messages.push_back(((Message*)vals[i].implementation())->getMQMessage());
		}
		if (params.size() == 1){
			return this->send(messages);
		}else{
			return this->send(messages, params[1]);
		}
	}else{

		Php::Value pvMessage = params[0];
		Message *message = (Message *)pvMessage.implementation();
		if (params.size() == 1){
			return this->send(message);
		}

		if (params.size() == 2){
			if (params[1].isBool()) {
				return this->send(message, (bool)params[1]);
			}
			if (params[1].isObject() && params[1].instanceOf("MessageQueue")) {
				return this->send(message, params[1]);
			}
		}
	}

	return false;
}


// virtual void sendOneway(MQMessage& msg, bool bSelectActiveBroker = false);
// virtual void sendOneway(MQMessage& msg, const MQMessageQueue& mq);
void Producer::sendOneway(Php::Parameters &params){
	Php::Value pvMessage = params[0];
	Message *message = (Message *)pvMessage.implementation();
	if (params.size() == 1){
		this->producer->sendOneway(message->getMQMessage());
	}else if (params.size() == 2 && params[1].isBool()){
		this->producer->sendOneway(message->getMQMessage(), (bool)params[1]);
	}else if (params.size() == 2 && params[1].instanceOf("MessageQueue")) {
		MessageQueue* messageQueue = (MessageQueue*)(params[1]).implementation();
		this->producer->sendOneway(message->getMQMessage(), messageQueue->getInstance());
	}
}


Php::Value Producer::getNamesrvAddr(){
	return this->producer->getNamesrvAddr();
}

void Producer::setSessionCredentials(Php::Parameters &param){
	std::string accessKey = param[0];
	std::string secretKey = param[1];
	std::string authChannel = param[2];

	this->producer->setSessionCredentials(accessKey, secretKey, authChannel);
}

Php::Value Producer::getSessionCredentials(){
	rocketmq::SessionCredentials sc = this->producer->getSessionCredentials();
	SessionCredentials *sessionCredentials = new SessionCredentials(&sc);
	Php::Value pv(Php::Object(SESSION_CREDENTIALS_CLASS_NAME , sessionCredentials));
	return pv;
}

void Producer::setNamesrvDomain(Php::Parameters &param){
	std::string domain = param[0];
	this->producer->setNamesrvDomain(domain);
}

Php::Value Producer::getNamesrvDomain(){
	return this->producer->getNamesrvDomain();
}

// int getRetryTimes() const;
Php::Value Producer::getRetryTimes(){
	return this->producer->getRetryTimes();
}

// void setRetryTimes(int times);
void Producer::setRetryTimes(Php::Parameters &param){
	this->producer->setRetryTimes(param[0]);
}


//int getSendMsgTimeout() const;
Php::Value Producer::getSendMsgTimeout(){
	return this->producer->getSendMsgTimeout();
}

//void setSendMsgTimeout(int sendMsgTimeout);
void Producer::setSendMsgTimeout(Php::Parameters &param){
	this->producer->setSendMsgTimeout(param[0]);
}

//int getCompressMsgBodyOverHowmuch() const;
Php::Value Producer::getCompressMsgBodyOverHowmuch(){
	return this->producer->getCompressMsgBodyOverHowmuch();
}

//void setCompressMsgBodyOverHowmuch(int compressMsgBodyOverHowmuch);
void Producer::setCompressMsgBodyOverHowmuch(Php::Parameters &param){
	this->producer->setCompressMsgBodyOverHowmuch(param[0]);
}

//int getCompressLevel() const;
Php::Value Producer::getCompressLevel(){
	return this->producer->getCompressLevel();
}

//void setCompressLevel(int compressLevel);
void Producer::setCompressLevel(Php::Parameters &param){
	this->producer->setCompressLevel(param[0]);
}

//int getMaxMessageSize() const;
Php::Value Producer::getMaxMessageSize(){
	return this->producer->getMaxMessageSize();
}

//void setMaxMessageSize(int maxMessageSize);
void Producer::setMaxMessageSize(Php::Parameters &param){
	this->producer->setMaxMessageSize(param[0]);
}

// void setTcpTransportPullThreadNum(int num);
void Producer::setTcpTransportPullThreadNum(Php::Parameters &param){
	this->producer->setTcpTransportPullThreadNum((int64_t)param[0]);
}

// const int getTcpTransportPullThreadNum() const;
Php::Value Producer::getTcpTransportPullThreadNum(){
	return (int64_t)this->producer->getTcpTransportPullThreadNum();
}

// void setTcpTransportConnectTimeout(uint64_t timeout);  // ms
void Producer::setTcpTransportConnectTimeout(Php::Parameters &param){
	this->producer->setTcpTransportConnectTimeout((int64_t)param[0]);
}
// const uint64_t getTcpTransportConnectTimeout() const;
Php::Value Producer::getTcpTransportConnectTimeout(){
	return (int64_t)this->producer->getTcpTransportConnectTimeout();
}

// void setTcpTransportTryLockTimeout(uint64_t timeout);  // ms
void Producer::setTcpTransportTryLockTimeout(Php::Parameters &param){
	this->producer->setTcpTransportTryLockTimeout((int64_t)param[0]);
}

// const uint64_t getTcpTransportConnectTimeout() const;
Php::Value Producer::getTcpTransportTryLockTimeout(){
	return (int64_t)this->producer->getTcpTransportTryLockTimeout();
}

//void setUnitName(std::string unitName);
void Producer::setUnitName(Php::Parameters &param){
	this->producer->setUnitName(param[0]);
}
//const std::string& getUnitName();
Php::Value Producer::getUnitName(){
	return this->producer->getUnitName();
}

void Producer::setLogLevel(Php::Parameters &param){
	this->producer->setLogLevel(rocketmq::elogLevel((int)param[0]));
}

Php::Value Producer::getLogLevel(){
	return this->producer->getLogLevel();
}

void Producer::setLogFileSizeAndNum(Php::Parameters &param){
	this->producer->setLogFileSizeAndNum(param[0], param[1]);
}

void registerProducer(Php::Namespace &rocketMQNamespace){
	Php::Class<Producer> producerClass("Producer");

	producerClass.method<&Producer::__construct>("__construct", { Php::ByVal("groupName", Php::Type::String), });
	producerClass.method<&Producer::__destruct>("__destruct");

	producerClass.method<&Producer::getInstanceName>("getInstanceName");
	producerClass.method<&Producer::setInstanceName>("setInstanceName", { Php::ByVal("groupName", Php::Type::String), });

	producerClass.method<&Producer::getNamesrvAddr>("getNamesrvAddr");
	producerClass.method<&Producer::setNamesrvAddr>("setNamesrvAddr", { Php::ByVal("nameserver", Php::Type::String), });

	producerClass.method<&Producer::setNamesrvDomain>("setNamesrvDomain", { Php::ByVal("domain", Php::Type::String), });
	producerClass.method<&Producer::getNamesrvDomain>("getNamesrvDomain");

	producerClass.method<&Producer::getGroupName>("getGroupName");
	producerClass.method<&Producer::setGroupName>("setGroupName", { Php::ByVal("groupName", Php::Type::String), });

	producerClass.method<&Producer::start>("start");
	producerClass.method<&Producer::shutdown>("shutdown");

	producerClass.method<&Producer::send>("send", { Php::ByVal("message", Php::Type::Null), Php::ByVal("queue", Php::Type::Null, false), });
	producerClass.method<&Producer::sendOneway>("sendOneway", { Php::ByVal("message", MESSAGE_CLASS_NAME), Php::ByVal("bSelectActiveBroker", Php::Type::Null, false) ,});

	producerClass.method<&Producer::getSessionCredentials>("getSessionCredentials");
	producerClass.method<&Producer::setSessionCredentials>("setSessionCredentials", {
			Php::ByVal("accessKey", Php::Type::String),
			Php::ByVal("secretKey", Php::Type::String),
			Php::ByVal("authChannel", Php::Type::String),
			});

	producerClass.method<&Producer::setRetryTimes>("setRetryTimes", { Php::ByVal("retryTimes", Php::Type::Numeric), });
	producerClass.method<&Producer::getRetryTimes>("getRetryTimes");

	producerClass.method<&Producer::getSendMsgTimeout>("getSendMsgTimeout");
	producerClass.method<&Producer::setSendMsgTimeout>("setSendMsgTimeout", {Php::ByVal("sendMsgTimeout", Php::Type::Numeric),});

	producerClass.method<&Producer::getCompressMsgBodyOverHowmuch>("getCompressMsgBodyOverHowmuch");
	producerClass.method<&Producer::setCompressMsgBodyOverHowmuch>("setCompressMsgBodyOverHowmuch", {Php::ByVal("compressMsgBodyOverHowmuch", Php::Type::Numeric),});

	producerClass.method<&Producer::getCompressLevel>("getCompressLevel");
	producerClass.method<&Producer::setCompressLevel>("setCompressLevel", {Php::ByVal("compressLevel", Php::Type::Numeric),});

	producerClass.method<&Producer::getMaxMessageSize>("getMaxMessageSize");
	producerClass.method<&Producer::setMaxMessageSize>("setMaxMessageSize", {Php::ByVal("maxMessageSize", Php::Type::Numeric),});

	producerClass.method<&Producer::getTcpTransportTryLockTimeout>("getTcpTransportTryLockTimeout");
	producerClass.method<&Producer::setTcpTransportTryLockTimeout>("setTcpTransportTryLockTimeout",{ Php::ByVal("timeout", Php::Type::Numeric), });

	producerClass.method<&Producer::getTcpTransportConnectTimeout>("getTcpTransportConnectTimeout");
	producerClass.method<&Producer::setTcpTransportConnectTimeout>("setTcpTransportConnectTimeout", {Php::ByVal("timeout", Php::Type::Numeric), });

	producerClass.method<&Producer::getTcpTransportPullThreadNum>("getTcpTransportPullThreadNum", {Php::ByVal("threadNum", Php::Type::Numeric), });
	producerClass.method<&Producer::setTcpTransportPullThreadNum>("setTcpTransportPullThreadNum", {Php::ByVal("threadNum", Php::Type::Numeric), });

	producerClass.method<&Producer::getUnitName>("getUnitName");
	producerClass.method<&Producer::setUnitName>("setUnitName", {Php::ByVal("unitName", Php::Type::String),});

	producerClass.method<&Producer::setLogLevel>("setLogLevel", {Php::ByVal("inputLevel", Php::Type::Numeric),});
	producerClass.method<&Producer::getLogLevel>("getLogLevel");
	producerClass.method<&Producer::setLogFileSizeAndNum>("setLogFileSizeAndNum", {Php::ByVal("fileNum", Php::Type::Numeric),Php::ByVal("perFileSize", Php::Type::Numeric),});

	producerClass.method<&Producer::setMessageTrace>("setMessageTrace", {Php::ByVal("messageTrace", Php::Type::Bool), });
	producerClass.method<&Producer::getMessageTrace>("getMessageTrace");

	producerClass.method<&Producer::getRetryTimes4Async>("getRetryTimes4Async");
	producerClass.method<&Producer::setRetryTimes4Async>("setRetryTimes4Async", {Php::ByVal("times", Php::Type::Numeric), });

	producerClass.method<&Producer::setLogPath>("setLogPath", {Php::ByVal("logPath", Php::Type::String), });

	producerClass.method<&Producer::getNameSpace>("getNameSpace");
	producerClass.method<&Producer::setNameSpace>("setNameSpace", {Php::ByVal("nameSpace", Php::Type::String), });

	producerClass.method<&Producer::version>("version");

	rocketMQNamespace.add(producerClass);
}


